internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,274 @@
+	tunnelID, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	// We use the same expiry for the password and the state entry.
+	// The state's expiry is used to clean up any dangling requests.
+	// The password expiry is used to invalidate old passwords.
+	now := tt.clock.Now()
+
+	password, err := tt.authn.generatePassword(tunnelID.String(), now)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKey, publicKey, err := tt.generateEphemeralSSHKey()
+	if err != nil {
+		return nil, err
+	}
+
+	args := state.SSHConnRequestArg{
+		TunnelID:            tunnelID.String(),
+		ModelUUID:           req.ModelUUID,
+		MachineId:           req.MachineID,
+		Expires:             now.Add(maxTimeout),
+		Username:            reverseTunnelUser,
+		Password:            password,
+		ControllerAddresses: tt.controller.Addresses(),
+		UnitPort:            0, // Allow the unit worker to determine the port.
+		EphemeralPublicKey:  publicKey.Marshal(),
+	}
+
+	err = tt.state.InsertSSHConnRequest(args)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanup := func() {
+		tt.delete(tunnelID.String())
+	}
+	// Make sure to use an unbuffered channel to ensure someone always
+	// has responsibility of the connection passed around.
+	tunnelReq := &Request{
+		recv:       make(chan (net.Conn)),
+		privateKey: privateKey,
+		cleanup:    cleanup,
+		dialer:     tt.dialer,
+	}
+
+	tt.add(tunnelID.String(), tunnelReq)
+
+	return tunnelReq, nil
+}
+
+func (tt *Tracker) add(tunnelID string, req *Request) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	tt.tracker[tunnelID] = req
+}
+
+func (tt *Tracker) get(tunnelID string) (*Request, bool) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	req, ok := tt.tracker[tunnelID]
+	return req, ok
+}
+
+func (tt *Tracker) delete(tunnelID string) {
+	tt.mu.Lock()
+	defer tt.mu.Unlock()
+	delete(tt.tracker, tunnelID)
+}
+
+// AuthenticateTunnel authenticates an SSH request for a tunnel.
+//
+// An SSH server is expected to call this method to validate that
+// the connection is a valid tunnel request.
+//
+// If the request is valid, the provided tunnelID should be
+// stored and provided alongside the network connection to PushTunnel.
+func (tt *Tracker) AuthenticateTunnel(username, password string) (tunnelID string, err error) {
+	if username != reverseTunnelUser {
+		return "", errors.New("invalid username")
+	}
+
+	return tt.authn.validatePassword(password)
+}
+
+// PushTunnel publishes a network connection for a tunnel.
+// This method should only be called after AuthenticateTunnel
+// which will provide the tunnelID.
+//
+// If the tunnelID is not valid an error is returned and the
+// caller should close the connection.
+//
+// This method blocks until a consumer runs Wait() on the
+// appropriate tunnel request. Use context.WithTimeout to control
+// the maximum time to wait.
+func (tt *Tracker) PushTunnel(ctx context.Context, tunnelID string, conn net.Conn) error {
+	req, ok := tt.get(tunnelID)
```

Comment:
Don't we need to close the `conn` if the tunnel does not exist or the context is cancelled?